source/puppet/5.0/lang_data_hash.md

```diff
@@ -29,17 +29,25 @@ notice($result)
 
 ## Syntax
 
-Hashes are written as a pair of curly braces containing any number of key/value pairs. A key is separated from its value by a `=>` (arrow, fat comma, or hash rocket), and adjacent pairs are separated by commas. An optional trailing comma is allowed between the final value and the closing curly brace.
+Hashes are written as a pair of curly braces containing any number of key/value pairs. Keys should be quoted with single or double quotes. A key is separated from its value by a `=>` (arrow, fat comma, or hash rocket), and adjacent pairs are separated by commas. An optional trailing comma is allowed between the final value and the closing curly brace.
 
 ``` puppet
-{ key1 => 'val1', key2 => 'val2' }
+{ 'key1' => 'val1', key2 => 'val2' }
 # Equivalent:
-{ key1 => 'val1', key2 => 'val2', }
+{ 'key1' => 'val1', key2 => 'val2', }
 ```
 
-Hash keys can be any data type, but you should generally only use strings or numbers.
+Hash keys can be any data type, but you should generally use only strings. You should not assign a hash with non-string keys to a resource attribute or class parameter, because Puppet cannot serialize non-string hash keys into the catalog.
```

Comment:
Here it could be pointed out that `you should generally use only strings (i.e. values that are in single or double quotes).`

Reply:
But depends on how the issue above was phrased.